src/System.CommandLine/Binding/ArgumentConverter.cs

```diff
@@ -106,59 +106,87 @@ internal static class ArgumentConverter
-            IReadOnlyCollection<string> tokens,
+            IReadOnlyList<string> tokens,
             ArgumentResult? argumentResult = null)
         {
-            if (type is null)
-            {
-                throw new ArgumentNullException(nameof(type));
-            }
-
-            if (tokens is null)
-            {
-                throw new ArgumentNullException(nameof(tokens));
-            }
-
             var itemType = type == typeof(string)
                                ? typeof(string)
                                : GetItemTypeIfEnumerable(type);
 
-            var parseResults = tokens
-                               .Select(arg => ConvertString(argument, itemType, arg))
-                               .ToArray();
-
-            var list = (IList) Activator.CreateInstance(typeof(List<>).MakeGenericType(itemType));
+            var (values, isArray) = type.IsArray
+                             ? (CreateArray(itemType!, tokens.Count), true)
+                             : (CreateList(itemType!, tokens.Count), false);
 
-            for (var i = 0; i < parseResults.Length; i++)
+            for (var i = 0; i < tokens.Count; i++)
             {
-                var result = parseResults[i];
+                var token = tokens[i];
+
+                var result = ConvertString(argument, itemType, token);
 
                 switch (result)
                 {
                     case FailedArgumentTypeConversionResult _:
                     case FailedArgumentConversionResult _:
                         if (argumentResult is { })
-                        {
+                        { 
                             argumentResult.OnlyTake(i);
-                        
+
                             // exit the for loop
-                            i = parseResults.Length;
+                            i = tokens.Count;
                             break;
                         }
 
                         return result;
-               
+
                     case SuccessfulArgumentConversionResult success:
-                        list.Add(success.Value);
+                        if (isArray)
+                        {
+                            values[i] = success.Value;
+                        }
+                        else
+                        {
+                            values.Add(success.Value);
+                        }
+
                         break;
                 }
             }
 
-            var value = type.IsArray
-                            ? (object) Enumerable.ToArray((dynamic) list)
-                            : list;
+            return Success(argument, values);
 
-            return Success(argument, value);
+            static IList CreateList(Type itemType, int capacity)
+            {
+                if (itemType == typeof(string))
+                {
+                    return new List<string>(capacity);
+                }
+                else if (itemType == typeof(int))
```

Comment:
Why are we optimizing for int?

Reply:
It's an experiment. `Activator.CreateInstance` sticks out as a perf problem, so I was trying to see if there are ways to reduce that hit. Of course this approach is optimized for a specific and maybe not common use case, so we can't really extend this to many types and maybe shouldn't bother at all.

Other ideas?